File: Difficulty: Medium/Largest subarray with 0 sum/largest-subarray-with-0-sum.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
/*You are required to complete this function*/

class Solution {
  public:
    int maxLen(vector<int>& arr, int n) {
        // Your code here
        unordered_map<int,int> mpp; //<sum,idx>
        int sum=0;
        int maxLen=0;
        
        for(int i=0;i<n;i++){
            sum+=arr[i];
            if(sum==0){ //if subarray starting from 0 idx.
                maxLen=max(maxLen,i+1);
            }
            else{
                if(mpp.find(sum)!=mpp.end()){
                    maxLen=max(maxLen,i-mpp[sum]);
                }
                else{
                    mpp[sum]=i;
                }
            }
        }
        return maxLen;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int m;
        cin >> m;
        vector<int> array1(m);
        for (int i = 0; i < m; ++i) {
            cin >> array1[i];
        }
        Solution ob;
        cout << ob.maxLen(array1, m) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Largest subarray with 0 sum/largest-subarray-with-0-sum.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Geeks {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine()); // Number of test cases
        for (int g = 0; g < t; g++) {
            // Read input array as a string, split by space, and convert to integers
            String[] str = br.readLine().trim().split(" ");
            int arr[] = new int[str.length];
            for (int i = 0; i < str.length; i++) {
                arr[i] = Integer.parseInt(str[i]);
            }
            // Print the result from maxLen function
            System.out.println(new Solution().maxLen(arr));
            System.out.println("~");
        }
    }
}

// } Driver Code Ends



class Solution {
    int maxLen(int nums[]) {
        // code here
        int n=nums.length;
        int maxLen=0;

        HashMap<Integer,Integer> mpp = new HashMap<>(); //map<prefixSum,idx>
        int prefixSum=0;

        mpp.put(0,-1);

        for(int i=0;i<n;i++){
            prefixSum+=nums[i];

            if(mpp.containsKey(prefixSum-0)){
                maxLen = Math.max(maxLen,i-mpp.get(prefixSum-0));
                continue;
            }

            mpp.put(prefixSum,i);
        }

        return maxLen;
    }
}